File: tobkiri_runtime/ecosystem/defaultspack/domain/ai_client/thinking_control.py
from __future__ import annotations

import re
from copy import deepcopy
from decimal import Decimal, InvalidOperation
from typing import Any


LEGACY_THINKING_LEVELS = ("none", "low", "medium", "high", "xhigh")
_NUMERIC_INPUT = re.compile(r"^(?P<number>(?:\d+(?:\.\d*)?|\.\d+))(?P<suffix>[kKmMbB]?)$")
_SAFE_BINDING_SEGMENT = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_ALLOWED_BINDING_ROOTS = {
    "reasoning",
    "reasoning_effort",
    "thinking",
    "thinking_budget",
    "thinking_level",
}
_SI_MULTIPLIERS = {
    "": Decimal(1),
    "k": Decimal(1_000),
    "m": Decimal(1_000_000),
    "b": Decimal(1_000_000_000),
}


def parse_numeric_shorthand(raw_value: Any) -> int | float:
    """Parse a finite decimal value with an optional decimal SI suffix."""
    text = str(raw_value or "").strip()
    match = _NUMERIC_INPUT.fullmatch(text)
    if match is None:
        raise ValueError("value must be a number with an optional k, m, or b suffix")
    try:
        value = Decimal(match.group("number")) * _SI_MULTIPLIERS[match.group("suffix").lower()]
    except (InvalidOperation, KeyError) as exc:
        raise ValueError("value must be a finite decimal number") from exc
    if not value.is_finite():
        raise ValueError("value must be a finite decimal number")
    integral = value.to_integral_value()
    return int(integral) if value == integral else float(value)


def normalize_thinking_control(profile: dict[str, Any] | None) -> dict[str, Any]:
    """Return the profile-owned thinking contract with legacy enum fallback."""
    source = profile if isinstance(profile, dict) else {}
    thinking = source.get("thinking") if isinstance(source.get("thinking"), dict) else {}
    explicit = source.get("thinking_control")
    if not isinstance(explicit, dict):
        explicit = thinking.get("control")
    if isinstance(explicit, dict):
        contract = deepcopy(explicit)
        contract["supported"] = contract.get("supported", True) is not False
        schema = contract.get("input_schema")
        contract["input_schema"] = deepcopy(schema) if isinstance(schema, dict) else {}
        binding = contract.get("request_binding")
        contract["request_binding"] = deepcopy(binding) if isinstance(binding, dict) else {}
        contract["source"] = "profile"
        return contract

    levels = thinking.get("levels")
    if not isinstance(levels, list):
        levels = source.get("thinking_levels")
    normalized_levels = (
        [str(level).strip() for level in levels if str(level or "").strip()]
        if isinstance(levels, list)
        else []
    )
    if not normalized_levels:
        normalized_levels = list(LEGACY_THINKING_LEVELS)
    return {
        "supported": bool(
            source.get("supports_thinking") or thinking.get("supported") or normalized_levels
        ),
        "input_schema": {"type": "enum", "values": normalized_levels},
        "request_binding": {},
        "source": "legacy",
    }


def validate_thinking_control(contract: dict[str, Any], raw_value: Any) -> dict[str, Any]:
    """Validate and normalize a raw thinking value against a profile contract."""
    raw = str(raw_value or "").strip()
    if contract.get("supported") is False:
        return _invalid(raw, "thinking control is not supported by this profile")
    schema = contract.get("input_schema")
    if not isinstance(schema, dict):
        return _invalid(raw, "thinking control input_schema is required")
    input_type = str(schema.get("type") or "").strip().lower()
    if raw == "auto":
        if schema.get("allow_auto") is not True:
            return _invalid(raw, "auto is not supported by this profile", input_type)
        if "auto_value" not in schema:
            return _invalid(raw, "profile does not declare an auto value", input_type)
        raw_for_validation = schema.get("auto_value")
    else:
        raw_for_validation = raw

    if input_type == "number":
        return _validate_number(schema, raw, raw_for_validation)
    if input_type == "enum":
        values = [
            str(value) for value in schema.get("values", []) if isinstance(value, (str, int, float))
        ]
        aliases = schema.get("aliases") if isinstance(schema.get("aliases"), dict) else {}
        normalized = aliases.get(raw_for_validation, raw_for_validation)
        if str(normalized) not in values:
            return _invalid(
                raw,
                f"value must be one of {', '.join(values)}",
                input_type,
            )
        return _valid(raw, str(normalized), input_type)
    if input_type == "text":
        normalized = str(raw_for_validation)
        maximum = _integer_bound(schema.get("max_length"), default=64)
        if not normalized or len(normalized) > maximum:
            return _invalid(
                raw,
                f"value must contain between 1 and {maximum} characters",
                input_type,
            )
        pattern = str(schema.get("pattern") or "").strip()
        if pattern:
            if len(pattern) > 256:
                return _invalid(raw, "profile thinking pattern is too long", input_type)
            try:
                matches = re.fullmatch(pattern, normalized)
            except re.error:
                return _invalid(raw, "profile thinking pattern is invalid", input_type)
            if matches is None:
                return _invalid(raw, "value does not match the profile pattern", input_type)
        return _valid(raw, normalized, input_type)
    return _invalid(raw, "thinking control type must be number, enum, or text")


def serialize_thinking_control(contract: dict[str, Any], normalized_value: Any) -> dict[str, Any]:
    """Serialize a normalized value using a safe, profile-declared request path."""
    binding = contract.get("request_binding")
    if not isinstance(binding, dict) or not binding:
        return {}
    path = str(binding.get("path") or "").strip()
    segments = path.split(".") if path else []
    if (
        not segments
        or segments[0] not in _ALLOWED_BINDING_ROOTS
        or len(segments) > 6
        or any(_SAFE_BINDING_SEGMENT.fullmatch(segment) is None for segment in segments)
    ):
        raise ValueError("thinking request_binding path is not allowed")
    template = binding.get("value", "$input")
    if template != "$input":
        raise ValueError("thinking request_binding value must be $input")
    value = normalized_value
    result: dict[str, Any] = {}
    cursor = result
    for segment in segments[:-1]:
        nested: dict[str, Any] = {}
        cursor[segment] = nested
        cursor = nested
    cursor[segments[-1]] = value
    return result


def _validate_number(schema: dict[str, Any], raw: str, raw_for_validation: Any) -> dict[str, Any]:
    try:
        normalized = parse_numeric_shorthand(raw_for_validation)
    except ValueError as exc:
        return _invalid(raw, str(exc), "number")
    value = Decimal(str(normalized))
    minimum = _decimal_bound(schema.get("min"))
    maximum = _decimal_bound(schema.get("max"))
    if minimum is not None and value < minimum:
        return _invalid(raw, f"value must be at least {minimum}", "number")
    if maximum is not None and value > maximum:
        return _invalid(raw, f"value must be at most {maximum}", "number")
    step = _decimal_bound(schema.get("step"))
    origin = minimum or Decimal(0)
    if step is not None and step > 0 and (value - origin) % step != 0:
        return _invalid(raw, f"value must use step {step}", "number")
    return _valid(raw, normalized, "number", unit=str(schema.get("unit") or ""))


def _decimal_bound(value: Any) -> Decimal | None:
    if value is None or isinstance(value, bool):
        return None
    try:
        result = Decimal(str(value))
    except InvalidOperation:
        return None
    return result if result.is_finite() else None


def _integer_bound(value: Any, *, default: int) -> int:
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        return default
    return parsed if parsed > 0 else default


def _valid(
    raw: str,
    normalized: Any,
    input_type: str,
    *,
    unit: str = "",
) -> dict[str, Any]:
    return {
        "valid": True,
        "raw": raw,
        "normalized": normalized,
        "input_type": input_type,
        "unit": unit,
        "message": "",
    }


def _invalid(raw: str, message: str, input_type: str = "") -> dict[str, Any]:
    return {
        "valid": False,
        "raw": raw,
        "normalized": None,
        "input_type": input_type,
        "unit": "",
        "message": message,
    }

File: tobkiri_runtime/ecosystem/defaultspack/webapp/src/features/models/thinkingControl.ts
import type { ModelProfile, ThinkingControlContract } from "../../lib/api";

export type ThinkingControlMode = "none" | "number" | "enum" | "text";

export function thinkingControlForProfile(profile?: ModelProfile | null): ThinkingControlContract {
  if (profile?.thinking_control?.source !== "legacy" && profile?.thinking_control?.input_schema) {
    return profile.thinking_control;
  }
  const values = profile?.thinking_levels?.length
    ? profile.thinking_levels
    : ["low", "medium", "high"];
  return {
    supported: Boolean(profile?.supports_thinking),
    input_schema: { type: "enum", values },
    request_binding: {},
    source: "legacy",
  };
}

export function thinkingControlMode(profile?: ModelProfile | null): ThinkingControlMode {
  const contract = thinkingControlForProfile(profile);
  if (!contract.supported) return "none";
  const type = contract.input_schema?.type;
  return type === "number" || type === "enum" || type === "text" ? type : "none";
}

export function thinkingControlCandidates(profile?: ModelProfile | null): string[] {
  const contract = thinkingControlForProfile(profile);
  if (contract.input_schema?.type !== "enum") return [];
  return (contract.input_schema.values ?? []).map(String).filter(Boolean);
}

export function thinkingControlInputError(
  profile: ModelProfile | null | undefined,
  value: string,
): string | null {
  const contract = thinkingControlForProfile(profile);
  const schema = contract.input_schema;
  const raw = value.trim();
  if (schema?.allow_auto && raw === "auto") return null;
  if (schema?.type === "enum") {
    return thinkingControlCandidates(profile).includes(raw)
      ? null
      : "Select a value declared by this model profile.";
  }
  if (schema?.type === "number") {
    if (!/^(?:\d+(?:\.\d*)?|\.\d+)[kKmMbB]?$/.test(raw)) {
      return "Enter a number with an optional k, m, or b suffix.";
    }
    const suffix = raw.slice(-1).toLowerCase();
    const multiplier = suffix === "k" ? 1_000 : suffix === "m" ? 1_000_000 : suffix === "b" ? 1_000_000_000 : 1;
    const numeric = Number(multiplier === 1 ? raw : raw.slice(0, -1)) * multiplier;
    if (!Number.isFinite(numeric)) return "Enter a finite number.";
    if (typeof schema.min === "number" && numeric < schema.min) return `Minimum: ${schema.min}`;
    if (typeof schema.max === "number" && numeric > schema.max) return `Maximum: ${schema.max}`;
    if (typeof schema.step === "number" && schema.step > 0) {
      const origin = typeof schema.min === "number" ? schema.min : 0;
      if (Math.abs((numeric - origin) % schema.step) > Number.EPSILON) return `Step: ${schema.step}`;
    }
    return null;
  }
  if (schema?.type === "text") {
    const maxLength = schema.max_length ?? 64;
    if (!raw || raw.length > maxLength) return `Enter 1–${maxLength} characters.`;
    if (schema.pattern) {
      try {
        if (!new RegExp(`^(?:${schema.pattern})$`).test(raw)) return "Value does not match this model profile.";
      } catch {
        return "This model profile declares an invalid pattern.";
      }
    }
    return null;
  }
  return "Thinking control is not supported by this model profile.";
}

export function normalizeThinkingControlInput(
  profile: ModelProfile | null | undefined,
  value: string,
): string | number {
  const schema = thinkingControlForProfile(profile).input_schema;
  const raw = value.trim();
  if (schema?.allow_auto && raw === "auto" && schema.auto_value !== undefined) {
    return schema.auto_value;
  }
  if (schema?.type !== "number") return raw;
  const suffix = raw.slice(-1).toLowerCase();
  const multiplier = suffix === "k" ? 1_000 : suffix === "m" ? 1_000_000 : suffix === "b" ? 1_000_000_000 : 1;
  return Number(multiplier === 1 ? raw : raw.slice(0, -1)) * multiplier;
}
